demo/lightning/checkpoint/multinode/fsspecfsdp.py

```diff
@@ -0,0 +1,107 @@
+                checkpoint.pop("optimizer_states", []))
+        })
+        save(converted_state, checkpoint_id=filepath, storage_writer=writer)
+
+        bucket = gcsfs.GCSFileSystem()
+        with bucket.open(os.path.join(filepath, _METADATA_FILENAME),
+                         'wb') as f:
+            torch.save(checkpoint, f)
+
+    def get_sharded_state_dict_context(
+            self, module: Module) -> Generator[None, None, None]:
+
+        from torch.distributed.fsdp.api import ShardedOptimStateDictConfig, ShardedStateDictConfig, StateDictType
+
+        state_dict_config = ShardedStateDictConfig(offload_to_cpu=True)
+        optim_state_dict_config = ShardedOptimStateDictConfig(
+            offload_to_cpu=True)
+        state_dict_type_context = FSDP.state_dict_type(
+            module=module,
+            state_dict_type=StateDictType.SHARDED_STATE_DICT,
+            state_dict_config=state_dict_config,
+            optim_state_dict_config=optim_state_dict_config,
+        )
+        return state_dict_type_context  # type: ignore[return-value]
+
+    def load_checkpoint(self, checkpoint_path):
+        # broadcast the path from rank 0 to ensure all the states are loaded from a common path
+        _ = Path(self.broadcast(checkpoint_path))
+
+        assert self.model is not None
+        assert self.lightning_module is not None
+
+        from torch.distributed.checkpoint.optimizer import load_sharded_optimizer_state_dict
+
+        state_dict_ctx = self.get_sharded_state_dict_context(self.model)
+        reader = FF.FsspecReader(checkpoint_path)
+
+        with state_dict_ctx:
+            module_state = {"model": self.model.state_dict()}
+            load(module_state, reader)
+            self.model.load_state_dict(
+                module_state["model"],
+                strict=self.lightning_module.strict_loading)
+
+            if self.lightning_module.trainer.state.fn == TrainerFn.FITTING and self.optimizers:
+
+                for idx, optim in enumerate(self.optimizers):
+                    optim_key = f"optimizer_{idx}"
+                    optim_state = load_sharded_optimizer_state_dict(
+                        model_state_dict=module_state["model"],
+                        optimizer_key=optim_key,
+                        storage_reader=reader,
+                    )
+                    flattened_osd = FSDP.optim_state_dict_to_load(
+                        optim_state_dict=optim_state[optim_key],
+                        model=self.model,
+                        optim=optim,
+                    )
+                    optim.load_state_dict(flattened_osd)
+
+        # Load metadata (anything not a module or optimizer)
+        new_path = os.path.join(checkpoint_path, _METADATA_FILENAME)
+        metadata = None
+        with reader.fs.create_stream(path=new_path,
+                                     mode='rb') as metadata_file:
+            metadata = torch.load(metadata_file)
+        return metadata
```

Comment:
Why does this function return just metadata? Looking at the pytorch FSDP implementation the load function seems to return the [full checkpoint](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/strategies/fsdp.py#L674).

Reply:
full checkpoint is being returned in case of no sharding, 

if it's sharded it returns just metadata https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/strategies/fsdp.py#L631

Reply:
That's right! But for sharded checkpoints it [returns just the metadata](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/strategies/fsdp.py#L631)